packages/flutter/lib/src/rendering/table.dart

```diff
@@ -586,124 +558,47 @@ class RenderTable extends RenderBox {
+    adoptChild(child);
+    if (after == null) {
+      // insert at the start.
+      _children.insert(0, child);
+    } else {
+      assert(_children.contains(after));
+      _children.insert(_children.indexOf(after) + 1, child);
     }
-    assert(cells != null);
-    assert(cells.length % columns == 0);
-    // fill a set with the cells that are moving (it's important not
-    // to dropChild a child that's remaining with us, because that
-    // would clear their parentData field)
-    final Set<RenderBox> lostChildren = HashSet<RenderBox>();
-    for (int y = 0; y < _rows; y += 1) {
-      for (int x = 0; x < _columns; x += 1) {
-        final int xyOld = x + y * _columns;
-        final int xyNew = x + y * columns;
-        if (_children[xyOld] != null && (x >= columns || xyNew >= cells.length || _children[xyOld] != cells[xyNew]))
-          lostChildren.add(_children[xyOld]!);
-      }
-    }
-    // adopt cells that are arriving, and cross cells that are just moving off our list of lostChildren
-    int y = 0;
-    while (y * columns < cells.length) {
-      for (int x = 0; x < columns; x += 1) {
-        final int xyNew = x + y * columns;
-        final int xyOld = x + y * _columns;
-        if (cells[xyNew] != null && (x >= _columns || y >= _rows || _children[xyOld] != cells[xyNew])) {
-          if (!lostChildren.remove(cells[xyNew]))
-            adoptChild(cells[xyNew]!);
-        }
-      }
-      y += 1;
-    }
-    // drop all the lost children
-    lostChildren.forEach(dropChild);
-    // update our internal values
-    _columns = columns;
-    _rows = cells.length ~/ columns;
-    _children = List<RenderBox?>.from(cells);
-    assert(_children.length == rows * columns);
-    markNeedsLayout();
-  }
-
-  /// Replaces the children of this table with the given cells.
-  void setChildren(List<List<RenderBox>>? cells) {
-    // TODO(ianh): Make this smarter, like setFlatChildren
-    if (cells == null) {
-      setFlatChildren(0, const <RenderBox?>[]);
-      return;
-    }
-    for (final RenderBox? oldChild in _children) {
-      if (oldChild != null)
-        dropChild(oldChild);
-    }
-    _children.clear();
-    _columns = cells.isNotEmpty ? cells.first.length : 0;
-    _rows = 0;
-    cells.forEach(addRow);
-    assert(_children.length == rows * columns);
   }
 
-  /// Adds a row to the end of the table.
+  /// Move the given `child` in the child list to be after another child.
   ///
-  /// The newly added children must not already have parents.
-  void addRow(List<RenderBox?> cells) {
-    assert(cells.length == columns);
-    assert(_children.length == rows * columns);
-    _rows += 1;
-    _children.addAll(cells);
-    for (final RenderBox? cell in cells) {
-      if (cell != null)
-        adoptChild(cell);
+  /// More efficient than removing and re-adding the child. Requires the child
+  /// to already be in the child list at some position. Pass null for `after` to
+  /// move the child to the start of the child list.
```

Comment:
it's more efficient, but it's still really expensive (potentially O(N) for the remove, O(N) for the insert, O(N) for the contains assert, O(N) for the indexOf...)